src/Lecture2.hs

```diff
@@ -264,8 +356,25 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
-
+eval _ (Lit i) = Right i
+eval vars (Var s) =
+  let
+    val = lookup s vars
+  in
+    if isJust val
+    then Right (fromJust val)
+    else Left (VariableNotFound s)
+eval vars (Add e1 e2) =
+  let
+    a = (eval vars e1)
+    b = (eval vars e2)
+  in -- the following feels more verbose than necessary. how to simplify?
+    if isLeft a 
+    then a
+    else if isLeft b
+         then b
+         else Right (fromRight 0 a + fromRight 0 b)
```

Comment:
Once you learn about monads, you'll be able to write this function even shorter! 🤓